cpp/src/gandiva/convert_timezone_holder.h

```diff
@@ -0,0 +1,282 @@
+        break;
+      case 5:
+        hours = parse_number(string_offset, 1, false);
+        minutes = parse_number(string_offset, 3, false);
+        seconds = 0;
+        break;
+      case 6:
+        hours = parse_number(string_offset, 1, false);
+        minutes = parse_number(string_offset, 4, true);
+        seconds = 0;
+        break;
+      case 7:
+        hours = parse_number(string_offset, 1, false);
+        minutes = parse_number(string_offset, 3, false);
+        seconds = parse_number(string_offset, 5, false);
+        break;
+      case 9:
+        hours = parse_number(string_offset, 1, false);
+        minutes = parse_number(string_offset, 4, true);
+        seconds = parse_number(string_offset, 7, true);
+        break;
+      default:
+        break;
+    }
+    char first = string_offset.at(0);
+    if (first != '+' && first != '-') {
+      throw "Invalid ID for ZoneOffset, plus/minus not found when expected: " +
+          string_offset;
+    }
+    if (first == '-') {
+      return seconds_from_minutes_hours(-hours, -minutes, -seconds);
+    } else {
+      return seconds_from_minutes_hours(hours, minutes, seconds);
+    }
+  }
+
+  void validate(int32_t hours, int32_t minutes, int32_t seconds) {
+    if (hours < -18 || hours > 18) {
+      throw "Zone offset hours not in valid range";
+    }
+    if (hours > 0) {
+      if (minutes < 0 || seconds < 0) {
+        throw "Zone offset minutes and seconds must be positive because hours is "
+              "positive";
+      }
+    } else if (hours < 0) {
+      if (minutes > 0 || seconds > 0) {
+        throw "Zone offset minutes and seconds must be negative because hours is "
+              "negative";
+      }
+    } else if ((minutes > 0 && seconds < 0) || (minutes < 0 && seconds > 0)) {
+      throw "Zone offset minutes and seconds must have the same sign";
+    }
+    if (minutes < -59 || minutes > 59) {
+      throw "Zone offset minutes not in valid range";
+    }
+    if (seconds < -59 || seconds > 59) {
+      throw "Zone offset seconds not in valid range";
+    }
+    if (abs(hours) == 18 && (minutes | seconds) != 0) {
+      throw "Zone offset not in valid range: -18:00 to +18:00";
+    }
+  }
+
+  std::chrono::seconds seconds_from_minutes_hours(int32_t hours, int32_t minutes,
+                                                  int32_t seconds) {
+    validate(hours, minutes, seconds);
+    int32_t total_seconds = hours * 3600 + minutes * 60 + seconds;
+    return std::chrono::seconds(total_seconds);
+  }
+
+  int32_t is_timezone_shift(const std::string& string_offset) {
+    if (string_offset.rfind("UTC") == 0 || string_offset.rfind("GMT") == 0) {
+      return 3;
+    } else if (string_offset.rfind("UT") == 0) {
+      return 2;
+    }
+    return 0;
+  }
+
+  std::shared_ptr<OffsetZone> src_offset_tz = NULLPTR;
+  std::shared_ptr<OffsetZone> dest_offset_tz = NULLPTR;
+
+  const time_zone* src_timezone = NULLPTR;
+  const time_zone* dest_timezone = NULLPTR;
```

Comment:
Could you use a shared pointer here? Because I think it is dangerous to use plain pointers as they may be not deallocated

Reply:
It's not possible because of the class used